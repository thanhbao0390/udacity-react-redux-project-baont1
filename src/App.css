/* .App {
  text-align: center;
}

.App-logo {
  height: 40vmin;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .App-logo {
    animation: App-logo-float infinite 3s ease-in-out;
  }
}

.App-header {
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
}

.App-link {
  color: rgb(112, 76, 182);
} */

@keyframes App-logo-float {
  0% {
    transform: translateY(0);
  }
  50% {
    transform: translateY(10px);
  }
  100% {
    transform: translateY(0px);
  }
}

.App {
  width: 70%;
  margin: auto;
  
}
.qs-add {
  border: 1px solid rgb(129, 155, 129);
  margin: 0px 0px 10px 0px;
  padding: 20px;
  background-color: rgb(227, 231, 227);
}

.qs-list {
  /* width: 70%; */
  margin: auto;
  
}

.qs-list-item {
  text-align: left;
  border: 1px solid rgb(129, 155, 129);
  margin: 0px 0px 10px 0px;
  padding: 20px;
  background-color: rgb(227, 231, 227);
  cursor: pointer;
}
.qs-item-title {
  font-size: larger;
}
.qs-item-author {
  font-size: medium;
  margin-bottom: 0px;
}
.qs-item-timestamp {
  font-size: small;
}
.qs-list-item-top {
  display: flex;
  justify-content: space-between;

}
.qs-item-option {
  margin: 20px;
}
.qs-tabs {
  
}
.qs-tabs-unanswered {
  display: inline-block;
  width: 50%;
  background-color: rgb(209, 248, 211);
  padding: 10px 0px;
  border: 0px;
  cursor: pointer;
}
.qs-tabs-unanswered-selected {
  font-weight: bold;
}
.qs-tabs-answered {
  display: inline-block;
  width: 50%;
  background-color: rgb(218, 226, 218);
  padding: 10px 0px;
  border: 0px;
  cursor: pointer;
}
.qs-tabs-answered-selected {
  font-weight: bold;
}

.qs-contents-unanswered {
  padding: 10px 10px 10px 10px;
  background-color: rgb(209, 248, 211);
}
.qs-contents-answered {
  padding: 10px 10px 10px 10px;
  background-color: rgb(218, 226, 218);
}

