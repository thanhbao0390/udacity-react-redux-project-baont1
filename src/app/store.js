import { configureStore } from '@reduxjs/toolkit';
import counterReducer from '../features/counter/counterSlice';
import questionReducer from '../features/question/questionSlice';
import { authentication } from './middleware/auth';

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    question: questionReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(authentication),
});
