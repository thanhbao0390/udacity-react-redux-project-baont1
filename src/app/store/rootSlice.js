import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { _saveQuestion, _saveQuestionAnswer } from '../api/_DATA';

let users = {
  sarahedo: {
    id: 'sarahedo',
    password: 'password123',
    name: 'Sarah Edo',
    avatarURL: 'sarahedo.jpg',
    answers: {
      "8xf0y6ziyjabvozdd253nd": 'optionOne',
      "6ni6ok3ym7mf1p33lnez": 'optionOne',
      "am8ehyc8byjqgar0jgpub9": 'optionTwo',
      "loxhs1bqm25b708cmbf3g": 'optionTwo'
    },
    questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
  },
  tylermcginnis: {
    id: 'tylermcginnis',
    password: 'abc321',
    name: 'Tyler McGinnis',
    avatarURL: 'tylermcginnis.jpg',
    answers: {
      "vthrdm985a262al8qx3do": 'optionOne',
      "xj352vofupe1dqz9emx13r": 'optionTwo',
    },
    questions: ['loxhs1bqm25b708cmbf3g', 'vthrdm985a262al8qx3do'],
  },
  mtsamis: {
    id: 'mtsamis',
    password: 'xyz123',
    name: 'Mike Tsamis',
    avatarURL: 'mtsamis.jpg',
    answers: {
      "xj352vofupe1dqz9emx13r": 'optionOne',
      "vthrdm985a262al8qx3do": 'optionTwo',
      "6ni6ok3ym7mf1p33lnez": 'optionOne'
    },
    questions: ['6ni6ok3ym7mf1p33lnez', 'xj352vofupe1dqz9emx13r'],
  },
  zoshikanlu: {
    id: 'zoshikanlu',
    password: 'pass246',
    name: 'Zenobia Oshikanlu',
    avatarURL: 'zoshikanlu.jpg',
    answers: {
      "xj352vofupe1dqz9emx13r": 'optionOne',
    },
    questions: [],
  }
}

let questions = {
  "8xf0y6ziyjabvozdd253nd": {
    id: '8xf0y6ziyjabvozdd253nd',
    author: 'sarahedo',
    timestamp: 1467166872634,
    optionOne: {
      votes: ['sarahedo'],
      text: 'Build our new application with Javascript',
    },
    optionTwo: {
      votes: [],
      text: 'Build our new application with Typescript'
    }
  },
  "6ni6ok3ym7mf1p33lnez": {
    id: '6ni6ok3ym7mf1p33lnez',
    author: 'mtsamis',
    timestamp: 1468479767190,
    optionOne: {
      votes: [],
      text: 'hire more frontend developers',
    },
    optionTwo: {
      votes: ['mtsamis', 'sarahedo'],
      text: 'hire more backend developers'
    }
  },
  "am8ehyc8byjqgar0jgpub9": {
    id: 'am8ehyc8byjqgar0jgpub9',
    author: 'sarahedo',
    timestamp: 1488579767190,
    optionOne: {
      votes: [],
      text: 'conduct a release retrospective 1 week after a release',
    },
    optionTwo: {
      votes: ['sarahedo'],
      text: 'conduct release retrospectives quarterly'
    }
  },
  "loxhs1bqm25b708cmbf3g": {
    id: 'loxhs1bqm25b708cmbf3g',
    author: 'tylermcginnis',
    timestamp: 1482579767190,
    optionOne: {
      votes: [],
      text: 'have code reviews conducted by peers',
    },
    optionTwo: {
      votes: ['sarahedo'],
      text: 'have code reviews conducted by managers'
    }
  },
  "vthrdm985a262al8qx3do": {
    id: 'vthrdm985a262al8qx3do',
    author: 'tylermcginnis',
    timestamp: 1489579767190,
    optionOne: {
      votes: ['tylermcginnis'],
      text: 'take a course on ReactJS',
    },
    optionTwo: {
      votes: ['mtsamis'],
      text: 'take a course on unit testing with Jest'
    }
  },
  "xj352vofupe1dqz9emx13r": {
    id: 'xj352vofupe1dqz9emx13r',
    author: 'mtsamis',
    timestamp: 1493579767190,
    optionOne: {
      votes: ['mtsamis', 'zoshikanlu'],
      text: 'deploy to production once every two weeks',
    },
    optionTwo: {
      votes: ['tylermcginnis'],
      text: 'deploy to production once every month'
    }
  },
}

const initialState = {
  questions: questions,
  userInfo: null,
  users: users,
  path: null,
  status: 'idle',
};

export const saveQuestionAsync = createAsyncThunk(
  'question/saveQuestion',
  async (question) => {
    const response = await _saveQuestion(question);
    return response;
  }
);

export const saveQuestionAnswerAsync = createAsyncThunk(
  'question/saveQuestionAnswer',
  async ({ authedUser, qid, answer }) => {
    const response = await _saveQuestionAnswer({ authedUser, qid, answer });
    return response;
  }
);

export const rootSlice = createSlice({
  name: 'root',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    logout: (state) => {
      state.userInfo = null;
    },
    login: (state, action) => {
      let user = state.users[action.payload.id];
      if (user) {
        state.userInfo = {
          ...user
        };
      } else {
        state.userInfo = null
      }
    },
    historyPath: (state, action) => {
      let path = action.payload.path;
      state.path = path;
    },

  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      .addCase(saveQuestionAsync.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(saveQuestionAsync.fulfilled, (state, action) => {
        state.status = 'idle';
        const authedUser = action.payload.author;
        state.users = {
          ...state.users,
          [authedUser]: {
            ...state.users[authedUser],
            questions: [
              ...state.users[authedUser].questions,
              action.payload.id
            ]
          }
        }
        state.questions = {
          ...state.questions,
          [action.payload.id]: action.payload
        }
      })
      .addCase(saveQuestionAsync.rejected, (state, action) => {
        state.status = 'error';
        state.error = action.error;
      })
      .addCase(saveQuestionAnswerAsync.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(saveQuestionAnswerAsync.fulfilled, (state, action) => {
        state.status = 'idle';
        const qid = action.meta.arg.qid;
        const answer = action.meta.arg.answer;
        const authedUser = action.meta.arg.authedUser;
        state.users = {
          ...state.users,
          [authedUser]: {
            ...state.users[authedUser],
            answers: {
              ...state.users[authedUser].answers,
              [qid]: answer
            }
          }
        }

        state.questions = {
          ...state.questions,
          [qid]: {
            ...state.questions[qid],
            [answer]: {
              ...state.questions[qid][answer],
              votes: state.questions[qid][answer].votes.concat([authedUser])
            }
          }
        }
      })
      .addCase(saveQuestionAnswerAsync.rejected, (state, action) => {
        state.status = 'error';
        state.error = action.error;
      });
  },
});

export const { logout, login, historyPath } = rootSlice.actions;

export default rootSlice.reducer;
