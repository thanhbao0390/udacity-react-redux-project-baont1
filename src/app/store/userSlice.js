import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

let users = {
  sarahedo: {
    id: 'sarahedo',
    password: 'password123',
    name: 'Sarah Edo',
    avatarURL: null,
    answers: {
      "8xf0y6ziyjabvozdd253nd": 'optionOne',
      "6ni6ok3ym7mf1p33lnez": 'optionOne',
      "am8ehyc8byjqgar0jgpub9": 'optionTwo',
      "loxhs1bqm25b708cmbf3g": 'optionTwo'
    },
    questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
  },
  tylermcginnis: {
    id: 'tylermcginnis',
    password: 'abc321',
    name: 'Tyler McGinnis',
    avatarURL: null,
    answers: {
      "vthrdm985a262al8qx3do": 'optionOne',
      "xj352vofupe1dqz9emx13r": 'optionTwo',
    },
    questions: ['loxhs1bqm25b708cmbf3g', 'vthrdm985a262al8qx3do'],
  },
  mtsamis: {
    id: 'mtsamis',
    password: 'xyz123',
    name: 'Mike Tsamis',
    avatarURL: null,
    answers: {
      "xj352vofupe1dqz9emx13r": 'optionOne',
      "vthrdm985a262al8qx3do": 'optionTwo',
      "6ni6ok3ym7mf1p33lnez": 'optionOne'
    },
    questions: ['6ni6ok3ym7mf1p33lnez', 'xj352vofupe1dqz9emx13r'],
  },
  zoshikanlu: {
    id: 'zoshikanlu',
    password: 'pass246',
    name: 'Zenobia Oshikanlu',
    avatarURL: null,
    answers: {
      "xj352vofupe1dqz9emx13r": 'optionOne',
    },
    questions: [],
  }
}
const userInfo = localStorage.getItem('userInfo')
  ? localStorage.getItem('userInfo')
  : null;

const initialState = {
  userInfo: userInfo,
  status: 'idle',
};

export const authAsync = createAsyncThunk(
  'auth/login',
  async (userInfo) => {
    return userInfo;
  }
);

export const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logout: (state) => {
      state.userInfo = null;

    },
    login: (state, action) => {
      let user = users[action.payload.id];
      if (user) {
        state.userInfo = {
          ...user
        };
      }
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(authAsync.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(authAsync.fulfilled, (state) => {
        state.status = 'idle';
      });
  },
});

export const usersList = () => {
  let usersList = [];
  Object.keys(users).forEach(function (key, index) {
    usersList.push(users[key]);
  });
  return usersList;
};

export const { logout, login } = userSlice.actions;

export default userSlice.reducer;
