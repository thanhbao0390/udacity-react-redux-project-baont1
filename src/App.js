import React from 'react';
// import logo from './logo.svg';
// import { Counter } from './features/counter/Counter';
import './App.css';
import { Question } from './features/question/Question';
import { QuestionAdd } from './features/question/QuestionAdd';
import { QuestionVote } from './features/question/QuestionVote';

import {
  createBrowserRouter,
  RouterProvider,
} from "react-router-dom";


function App() {

  const user = 'sarahedo';

  const router = createBrowserRouter([
    {
      path: "/",
      element: <Question user={user} />,
    },
    {
      path: "/add",
      element: <QuestionAdd user={user} />,
    },
    {
      path: "/vote/:id",
      element: <QuestionVote user={user} />,
    },
  ]);

  return (
    <div className="App">
      <RouterProvider router={router} />
      {/* <header className="App-header">
        <img src={logo} className="App-logo" alt="logo" />
        
        <Counter />
        <p>
          Edit <code>src/App.js</code> and save to reload.
        </p>
        <span>
          <span>Learn </span>
          <a
            className="App-link"
            href="https://reactjs.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux
          </a>
          <span>, </span>
          <a
            className="App-link"
            href="https://redux-toolkit.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            Redux Toolkit
          </a>
          ,<span> and </span>
          <a
            className="App-link"
            href="https://react-redux.js.org/"
            target="_blank"
            rel="noopener noreferrer"
          >
            React Redux
          </a>
        </span>
      </header> */}
    </div>
  );
}

export default App;
