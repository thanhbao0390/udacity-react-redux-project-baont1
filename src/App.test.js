import renderer from 'react-test-renderer';
import { _saveQuestion, _saveQuestionAnswer } from './app/api/_DATA';

import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { store } from './app/store';
import App from './App';

describe('screen login', () => {
    let loginOutput = `<body>
                        <div>
                        <div
                            class="App"
                        >
                            <div
                            class="login-contents"
                            >
                            <label>
                                Choose user to login:
                            </label>
                            <select>
                                <option
                                disabled=""
                                value=""
                                >
                                Move to...
                                </option>
                                <option
                                value="sarahedo"
                                >
                                Sarah Edo
                                </option>
                                <option
                                value="tylermcginnis"
                                >
                                Tyler McGinnis
                                </option>
                                <option
                                value="mtsamis"
                                >
                                Mike Tsamis
                                </option>
                                <option
                                value="zoshikanlu"
                                >
                                Zenobia Oshikanlu
                                </option>
                            </select>
                            </div>
                        </div>
                        </div>
                    </body>`
    it('renders correctly', () => {
        const tree = renderer
            .create(loginOutput)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('_saveQuestion', () => {
    it('will return formatted question if all data is passed correctly', async () => {
        var question = {
            author: 'sarahedo',
            optionOneText: 'take a Option 1',
            optionTwoText: 'take a Option 2',
        };
        var result = await _saveQuestion(question);
        expect(result.author).toEqual('sarahedo');
        expect(result.optionOne.text).toEqual('take a Option 1');
        expect(result.optionTwo.text).toEqual('take a Option 2');
    });

    it('will return error if data passed is incorrect', async () => {
        var wrongQuestion = {
            optionOne: {
                text: 'take a Option 1',
            }
        };
        await expect(_saveQuestion(wrongQuestion)).rejects.toEqual('Please provide optionOneText, optionTwoText, and author');
    });
})

describe('_saveQuestionAnswer', () => {
    it('will return true if all data is passed correctly', async () => {
        var info = {
            authedUser: 'zoshikanlu',
            qid: '8xf0y6ziyjabvozdd253nd',
            answer: 'optionTwo',
        };
        var result = await _saveQuestionAnswer(info);
        expect(result).toEqual(true);
    });

    it('will return error if data passed is incorrect', async () => {
        var wrongInfo = {
            authedUser: 'zoshikanlu',
            answer: 'optionTwo',
        };
        await expect(_saveQuestionAnswer(wrongInfo)).rejects.toEqual('Please provide authedUser, qid, and answer');
    });
})