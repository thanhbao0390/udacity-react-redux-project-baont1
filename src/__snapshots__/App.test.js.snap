// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screen login renders correctly 1`] = `
"<body>
                        <div>
                        <div
                            class=\\"App\\"
                        >
                            <div
                            class=\\"login-contents\\"
                            >
                            <label>
                                Choose user to login:
                            </label>
                            <select>
                                <option
                                disabled=\\"\\"
                                value=\\"\\"
                                >
                                Move to...
                                </option>
                                <option
                                value=\\"sarahedo\\"
                                >
                                Sarah Edo
                                </option>
                                <option
                                value=\\"tylermcginnis\\"
                                >
                                Tyler McGinnis
                                </option>
                                <option
                                value=\\"mtsamis\\"
                                >
                                Mike Tsamis
                                </option>
                                <option
                                value=\\"zoshikanlu\\"
                                >
                                Zenobia Oshikanlu
                                </option>
                            </select>
                            </div>
                        </div>
                        </div>
                    </body>"
`;
